File: python/tutorials/tour-of-orchestration-python/app/sagas/service.py
import restate
from typing import Callable, Awaitable, List

from app.utils import create_recurring_payment, remove_recurring_payment, create_subscription, remove_subscription
from app.types import SubscriptionRequest


subscription_saga = restate.Service("SubscriptionSaga")


@subscription_saga.handler()
async def add(ctx: restate.Context, req: SubscriptionRequest) -> None:
    compensations: List[Callable[[], Awaitable[None]]] = []

    try:
        payment_id = ctx.rand_uuid()

        # Add compensation for payment
        compensations.append(
            lambda: ctx.run_typed(
                "undo-pay", lambda: remove_recurring_payment(payment_id)
            )
        )

        # Create payment
        pay_ref = await ctx.run_typed(
            "pay", lambda: create_recurring_payment(req["creditCard"], payment_id)
        )

        # Process subscriptions
        for subscription in req["subscriptions"]:
            # Add compensation for this subscription
            compensations.append(
                lambda s=subscription: ctx.run_typed(
                    f"undo-{s}", lambda: remove_subscription(req["userId"], s)
                )
            )

            # Create subscription
            await ctx.run_typed(
                f"add-{subscription}",
                lambda s=subscription: create_subscription(req["userId"], s, pay_ref),
            )

    except restate.TerminalError as e:
        # Run compensations in reverse order
        for compensation in reversed(compensations):
            await compensation()
        raise e


app = restate.app(services=[subscription_saga])

File: kotlin/tutorials/tour-of-orchestration-kotlin/src/main/kotlin/my/example/sagas/service.kt
import dev.restate.sdk.Context
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.endpoint
import java.util.*

data class SubscriptionRequest(
    val userId: String,
    val creditCard: String,
    val subscriptions: Array<String>
)

// Mock function to create recurring payment
suspend fun createRecurringPayment(creditCard: String, paymentId: String): String {
    return "payRef-${UUID.randomUUID()}"
}

// Mock function to remove recurring payment
suspend fun removeRecurringPayment(paymentId: String) {
    println("Removing recurring payment: $paymentId")
}

// Mock function to create subscription
suspend fun createSubscription(userId: String, subscription: String, paymentRef: String) {
    println("Creating subscription for user: $userId, subscription: $subscription, paymentRef: $paymentRef")
}

// Mock function to remove subscription
suspend fun removeSubscription(userId: String, subscription: String) {
    println("Removing subscription for user: $userId, subscription: $subscription")
}

@Service
class SubscriptionSaga {

    @Handler
    suspend fun add(ctx: Context, req: SubscriptionRequest) {
        val compensations = mutableListOf<suspend () -> Unit>()
        
        try {
            val paymentId = ctx.random().nextUUID().toString()

            compensations.add {
                ctx.runBlock("undo-pay") { 
                    removeRecurringPayment(paymentId) 
                }
            }
            
            val payRef = ctx.runBlock(
                "pay",
                String::class
            ) {
                createRecurringPayment(req.creditCard, paymentId)
            }

            for (subscription in req.subscriptions) {
                compensations.add {
                    ctx.runBlock("undo-$subscription") {
                        removeSubscription(req.userId, subscription)
                    }
                }
                
                ctx.runBlock("add-$subscription") {
                    createSubscription(req.userId, subscription, payRef)
                }
            }
        } catch (e: TerminalException) {
            // Run compensations in reverse order
            for (compensation in compensations.reversed()) {
                compensation()
            }
            throw e
        }
    }
}

fun main() {
    RestateHttpServer.listen(endpoint {
        bind(SubscriptionSaga())
    })
}